### DISCUSSION TITLE
Why `cabal repl` can play faster than `cabal run`

### DISCUSSION THREAD

**Discussion by WendaoLee** [haskell/cabal#9517] [Q&A] [answered]
Or for a short: What things that `cabal run` need to do but `cabal repl` not to do?Why these things cost so much time?

---
I'm writing a Haskell programmer with `cabal` now.At the beginning,I always wrote some test cases in the `main` and used `cabal run` to check for it.But it always took about half-to-one minute on linking.Then I found `cabal repl` can play the same role,and I surprisingly found that `cabal repl` plays so faster than `cabal run`.According to the log in the terminal,it seems that `cabal repl` don't need `link`,it just compile and then go running.

That made me confused:Why `cabal repl` don't need `Link`?And what's the more that `cabal run` do?

I'm sorry it looks like a stupid question.But I'm really wonder the reason behind it.It smells so interesting. I'd appriciate if someone can give me any answer or reference.

**Comment by ulysses4ever** [comment]
For the same reason GHCi is faster to load a file than GHC to compile it: GHC's native backend is slow, but GHCi doesn't use the backend, it compiles to high-level bytecode (which it can interpret later on) rather than a native executable.

**Reply by WendaoLee** [reply]
Is the 'high-level bytecode' represented by the `.hi` file?

**Reply by Mikolaj** [reply]
No, another one. The `.hi` files are interface files. I think ghci doesn't write any files with the bytecode, which is yet another speedup. However, consult GHC documentation to make sure.

**Reply by fgaz** [reply]
@WendaoLee if you want to learn more about the general concepts, here are the respective wikipedia pages:

* https://en.wikipedia.org/wiki/Compiled_language
* https://en.wikipedia.org/wiki/Interpreter_(computing)

Haskell is both :)

**Reply by WendaoLee** [reply]
Ah, thanks. 😽 Though to be honest, I know these concepts(but I forgot `GHCI` interepted the program,thanks for your remind), and I was hoping someone might provide me with articles or references that delve deeper into the mechanisms behind these things when I asked this question... Well, I apologize if my expectations seemed a bit demanding and vexatious. I truly appreciate your help.

**Reply by ulysses4ever** [reply, marked as answer]
@WendaoLee it's all right. The thing is, you asked essentially a GHC question on a Cabal forum. You may have a better luck asking it on a GHC forum. For example, in the ghc-devs mailing list.

**Reply by WendaoLee** [reply]
Yes,you are right.😽Sorry for interruption.Thanks for your suggestion.
